### DISCUSSION TITLE
Synology Docker - failed to get urandom (RuntimeError)

### DISCUSSION THREAD

**Discussion by OxfordRich** [pglombardo/PasswordPusher#647] [Q&A]
Hi! I've been running PasswordPusher since early versions, without issue, in Docker on a Synology NAS. Today I went to update the image and I find that the container stops. I'm getting the folllwing error in the log: 

```
[1] ! Unable to load application: RuntimeError: failed to get urandom
bundler: failed to load command: puma (/opt/PasswordPusher/vendor/bundle/ruby/3.1.0/bin/puma)
/usr/local/lib/ruby/3.1.0/securerandom.rb:75:in `urandom': failed to get urandom (RuntimeError)
```
Is there an additional requirement or anything I can do to get newer builds running in Docker on my Synology, please?

**Comment by OxfordRich** [comment]
Looks like version 1.15.1 is the last version that will start on my NAS. Presumably this has something to do with Ruby 3.1.0 which is mentioned from 1.16.0 onwards.  I'm seeing a "rake aborted!" error.

**Reply by pglombardo** [reply]
Hi @OxfordRich,

> Unable to load application: RuntimeError: failed to get urandom

This is an odd error that I haven't seen and hasn't been reported previously.  Some googling says that running out of open file handles can do this.  I'd hate to ask you to reboot your NAS but that would tell us if it's an open file handle problem.  You can also check the configured limit with `ulimit -n` or increase it using `sysctl`.

The change in Ruby major version could definitely trigger this.

Could you take a look at that and let me know?

**Reply by OxfordRich** [reply]
Thanks so much for your swift response. I expect there aren't too many people running this via Docker on a Synology. 
I'm almost certain it's the Ruby version, as I'm able to switch back to older versions from 1.15.1 and backwards. The NAS doesn't do much else - although I'll try a reboot and report back if that does make any difference. The file handles are set to 1024 as is standard. 

I think I'll probably build a more easily updated docker host via VMware Photon and abandon the messing about with Synology.

**Reply by pglombardo** [reply]
Hi @OxfordRich - any updates on this?

In separate news, I was thinking of buying a Synology or alternative myself but haven't had the budget to burn lately.

**Reply by OxfordRich** [reply]
Hi, I'm afraid a reboot didn't make any difference. I now host your excellent project in VMware's Photon. 

I would be happy to get you remote access some time to confirm exactly what the missing requirement is, if you wanted. I think we already narrowed it to the Ruby version change not being supported in the current Synology OS - I'll try to remember to test it again following any Synology OS updates.

**Comment by MiranoVerhoef** [comment]
@pglombardo 

I was wondering if we could troubleshoot this furter.

I can provide you with a virtual DSM if needed

**Reply by pglombardo** [reply]
Hi @MiranoVerhoef - I found [this issue](https://github.com/docker-library/ruby/issues/383) which seems to be the same error/cause.

It points to an older kernel or Docker libraries as being the cause.

Testing with the following is a good way to easily reproduce/test:

```
docker run -it --rm ruby:3.1-bullseye bash
require 'securerandom'
```

Just a note that pwpush containers are now up to 3.2.

Could you take a look at that linked issue and investigate a bit?  I'm a bit underwater with tasks but don't want to leave you and @OxfordRich stranded without an answer.

**Reply by MiranoVerhoef** [reply]
Yeah we are trying to get the latest version (with ruby 2.3.0)

I will see if the issue above helps us!, thank you for answering

**Reply by OxfordRich** [reply]
Well, that's interesting. My situation matches the issue that @pglombardo linked entirely. 

```
`Richard@RS814:~$ sudo docker run -it --rm ruby:3.1-bullseye bash
root@fdf8c1f7d103:/# irb
irb(main):001:0> require securerandom
/usr/local/lib/ruby/3.1.0/securerandom.rb:75:in urandom: failed to get urandom (RuntimeError)
        from /usr/local/lib/ruby/3.1.0/securerandom.rb:75:in `singleton class
        from /usr/local/lib/ruby/3.1.0/securerandom.rb:42:in `<module:SecureRandom>
        from /usr/local/lib/ruby/3.1.0/securerandom.rb:41:in `<top (required)>
        from <internal:/usr/local/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require
        from <internal:/usr/local/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require
        from (irb):1:in `<main>
        from /usr/local/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>
        from /usr/local/bin/irb:25:in `load
        from /usr/local/bin/irb:25:in `<main>
```

but 

```
Richard@RS814:~$ sudo docker run -it --rm ruby:3.2-rc-slim bash
Unable to find image ruby:3.2-rc-slim locally
3.2-rc-slim: Pulling from library/ruby
3f4ca61aafcd: Pull complete
2c1347bf51c5: Pull complete
4a72f1f78ce7: Pull complete
c605cd2909ab: Pull complete
1fa1db826207: Pull complete
Digest: sha256:ba2d4dc84aa3b215f2fa04e88dd37de3ba135519679b9af33c2e7f5d72e23520
Status: Downloaded newer image for ruby:3.2-rc-slim
root@e638bb7326aa:/# irb
irb(main):002:0> require securerandom
=> true
```
and mine reports

```
Richard@RS814:~$ cat /proc/version
Linux version 3.10.108 (root@build3) (gcc version 8.5.0 (GCC) ) #42962 SMP Tue Oct 18 15:01:39 CST 2022
```
Would be interesting for me to try the very latest build of PasswordPusher back on the Synology now, just to see if moving to 3.2 has solved this. Will try that and report back.

**Reply by OxfordRich** [reply]
Laest version still fails on the Synology. 

I run this successfully via VMware Proton now, so don't worry too much about me! Pleased to see there is at least one other person using Docker on a Synology though.

```
Password Pusher: migrating database to latest...
rake aborted!
failed to get urandom
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/zeitwerk-2.6.6/lib/zeitwerk/loader/config.rb:89:in `initialize'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/zeitwerk-2.6.6/lib/zeitwerk/loader.rb:97:in `initialize'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/autoloaders.rb:22:in `new'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/autoloaders.rb:22:in `initialize'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/application.rb:121:in `new'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/application.rb:121:in `initialize'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/railtie.rb:184:in `new'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/railtie.rb:184:in `instance'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/application.rb:78:in `instance'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails.rb:40:in `application'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails.rb:64:in `root'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/config-4.1.0/lib/config/integrations/rails/railtie.rb:7:in `preload'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/railtie.rb:226:in `public_send'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/railtie.rb:226:in `method_missing'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/config-4.1.0/lib/config/integrations/rails/railtie.rb:21:in `block in <class:Railtie>'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:92:in `block in execute_hook'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:85:in `with_execution_control'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:90:in `execute_hook'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:76:in `block in run_load_hooks'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:75:in `each'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4.1/lib/active_support/lazy_load_hooks.rb:75:in `run_load_hooks'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/railties-7.0.4.1/lib/rails/application.rb:74:in `inherited'
/opt/PasswordPusher/config/application.rb:10:in `<module:PasswordPusher>'
/opt/PasswordPusher/config/application.rb:9:in `<top (required)>'
/opt/PasswordPusher/Rakefile:5:in `require'
/opt/PasswordPusher/Rakefile:5:in `<top (required)>'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli/exec.rb:58:in `load'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli/exec.rb:58:in `kernel_load'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli/exec.rb:23:in `run'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli.rb:486:in `exec'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli.rb:31:in `dispatch'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/cli.rb:25:in `start'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/exe/bundle:48:in `block in <top (required)>'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
/opt/PasswordPusher/vendor/bundle/ruby/3.2.0/gems/bundler-2.3.25/exe/bundle:36:in `<top (required)>'
/usr/local/bundle/bin/bundle:25:in `load'
/usr/local/bundle/bin/bundle:25:in `<main>'
(See full trace by running task with --trace)
```

**Reply by skydoy** [reply]
Any ideias, im having the same issue :(

**Reply by pglombardo** [reply]
HI @skydoy et. al. - I just took another look at this and unfortunately every avenue that I go down in other similar application repositories ends up at these two issues:

https://github.com/docker-library/redmine/issues/314 (last update last week)
https://github.com/docker-library/redmine/issues/264

Redmine is an application built with the same tech stack as Password Pusher.

@MiranoVerhoef or @OxfordRich - Did either of you ever make any progress on this issue?
